Pyomo. DoE: Sensitivity initialization

smondal13 opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes # .

## Summary/Motivation:
- Initializing the model to a previously solved point in the design space can help the model converge at another point in the design space for sensitivity analysis
- Better error messages in the method can help the user in debugging
- Saving the file can be helpful if the model takes a long time to solve or if the user wants it.
- The pattern generator in the utils.py file can be used in other cases where it can help with the initialization
- Correlation matrix computation functionality will be helpful in computing correlation among the parameters


## Changes proposed in this PR:
- Added a new method `compute_FIM_factorial()` in Pyomo.DoE for sensitivity initialization
- Added `snake_traversal_grid_sampling()` and `compute_correlation_matrix()` in utils.py

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

smondal13 committed:
added zigzag in utils and compute fim in doe.py

smondal13 committed:
Merge branch 'Pyomo:main' into sensitivity-initialization

smondal13 committed:
The compute_FIM_factorial() using zigzag pattern is now working as expected. Now, I need to add cases when some of the design variables are not changed.

smondal13 committed:
Merge branch 'sensitivity-initialization' of github.com:smondal13/pyomo into sensitivity-initialization

smondal13 committed:
Added utils.py as utils_updated.py from adding_eigen_values branch to use compute_FIM_metrics() method. Added both the nested for loop and change_one_design_at_a_time argument to compute_FIM_factorial() in doe.py. Everything is working fine.

smondal13 committed:
this version of compute_FIM_factorial() runs fine. The argument `design_values` does not need to be in the same order as the `experiment_inputs` as long as the key matches, we are good. Also, if we don't want to change a design variable, just not passing it as a key in the `design_values` dictionary will do the trick. Tested with the `reactor_example.py`

smondal13 committed:
Changed documentation and improved the error message

smondal13 committed:
Merge branch 'Pyomo:main' into sensitivity-initialization

smondal13 committed:
updated comments

adowling2 commented:
Does this PR depend on changes in #3575?

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1617,6 +1628,221 @@ def compute_FIM_full_factorial(
+        # design_map_keys so that design_point can be constructed correctly in the loop.
+        des_ranges = [design_values[k.name] for k in design_map_keys]
+        if change_one_design_at_a_time:
+            factorial_points = generate_snake_zigzag_pattern(*des_ranges)
```
Let's brainstorm different sensitivity analysis sequences we might use. We can then define an enum.

adowling2 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -100,3 +100,43 @@ def rescale_FIM(FIM, param_vals):
 #     return param_list
+
+
+def generate_snake_zigzag_pattern(*lists):
```
Let's come up with a better name. I do not think of "snake" as a technical term.

adowling2 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -100,3 +100,43 @@ def rescale_FIM(FIM, param_vals):
+                yield (value,) + sub_pattern
+
+    # Start the recursion at the first list (depth 0) with an initial sum of 0.
+    yield from _generate_recursive(depth=0, index_sum=0)
```
How are we going to test this function? IMO, the next step is to develop some tests for this. That way, as you try to run the sensitivity analysis for an example, you are least know this part of the code is correct.

smondal13 replied:
@adowling2 , I did check it manually before. I have recently added a test for this in `test_utils.py` for up to 3 variables, and also for different array-like structures. From the test, it works the way we want.

adowling2 reviewed (commented):
@smondal13 I left some initial comments.

smondal13 committed:
Added argument checking for generate zigzag pattern.

Also added test for patter generator

smondal13 committed:
changed the name of the pattern generator

smondal13 committed:
Merge branch 'adding_eigen_values' into sensitivity-initialization

smondal13 committed:
Merge branch 'Pyomo:main' into sensitivity-initialization

smondal13 committed:
Merge branch 'sensitivity-initialization' of github.com:smondal13/pyomo into sensitivity-initialization

smondal13 committed:
Changed the name of the grid sampling function and deleted the utils_updated.py file

smondal13 committed:
Merge remote-tracking branch 'sscini/add-update-model-utility' into sensitivity-initialization

smondal13 committed:
added rel_change and abs_change for change in design

smondal13 committed:
added update_suffix function in compute the sensitivity

smondal13 committed:
working code. added design vals and also changed the settings of pd df. The code is run by using a simple model.

smondal13 committed:
Merge branch 'Pyomo:main' into sensitivity-initialization

smondal13 commented:
@adowling2 , @djlaky This PR is ready for design review.

smondal13 committed:
changed some minor comments and docstrings

sscini committed:
updated with sscini's `utils.py`.

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Merge remote-tracking branch 'origin/main' into sensitivity-initialization

smondal13 committed:
Merge branch 'sensitivity-initialization' of github.com:smondal13/pyomo into sensitivity-initialization

smondal13 committed:
Merge branch 'Pyomo:main' into sensitivity-initialization

smondal13 committed:
added new func and changed test files

added a new correlation matrix computation in  doe/utils.py, added TODOs.

smondal13 committed:
changed if statement to ternary operator and minor changes in return statement

smondal13 committed:
Revert "changed if statement to ternary operator and minor changes in return statement"

This reverts commit a063c3384cfe44da9ba642bdf01e20314f7ae548.

smondal13 committed:
Reapply "changed if statement to ternary operator and minor changes in return statement"

This reverts commit 9af160d87615e3ff50e4f9bcc11150af75ce7a5c.

smondal13 committed:
deleted test_solve

smondal13 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1611,6 +1622,333 @@ def compute_FIM_full_factorial(
+        rel_change: list = None,
+        n_designs: int = 5,
+        method="sequential",
+        df_settings=(True, None, None, 500),
```
Should I use this tuple? I was reluctant to use 4 different arguments for the pd df setting / configuration

smondal13 committed:
added a strategy to compute the fractional factorial as well.

smondal13 commented:
TODO: Create a new suffix `design_var`s and pass that to Stephen's `update_model_from_suffix()`

smondal13 committed:
Created a temporary suffix `design_suff` to contain the keys of the user provided design variables

smondal13 committed:
Merge branch 'Pyomo:main' into sensitivity-initialization

smondal13 committed:
Merge remote-tracking branch 'upstream/main' into sensitivity-initialization

smondal13 committed:
changed the import location for `update_model_from_suffix` and resolved conflict of doe/utils.py

smondal13 committed:
cleaned doe/utils.py

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1617,6 +1628,221 @@ def compute_FIM_full_factorial(
 
         return self.fim_factorial_results
 
+    def compute_FIM_factorial(
```
Yes, a depreciation warning sounds reasonable.

jsiirola replied:
Actually, I am really confused: why did you create a new method *in the same file* that did effectively the same thing?  Does this replace the other method?  Is it a superset of the other method?  Do we need both?

Also, you might consider renaming this function: I don't know what a `FIM factorial` is (nor how to compute one).  Maybe something like `compute_FIM_from_full_factorial_sampling`?

smondal13 replied:
@jsiirola This function is a superset of the already existing `compute_FIM_full_factorial()` method.  So, I am not sure if I should keep it under the same name since the "full" in the method name can be misleading. It can use a full factorial (changing all the decision variables) and a fractional factorial (changing some of the decision variables). This method basically analyzes the sensitivity.

@adowling2, @slilonfe5, @snarasi2, @sscini  Can we name it something like `compute_sensitivity_at_design_points()`? Or similar to the one @jsiirola suggested, `compute_FIM_from_factorial_sampling`?

slilonfe5 replied:
@smondal13 If sensitivity is what you are doing, I think `compute_sensitivity_at_design_points` is better.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1802,6 +1815,350 @@ def compute_FIM_full_factorial(
+        file_name: str = None,
+    ):
+        """Will run a simulation-based factorial exploration of the experimental input
+        space (i.e., a ``grid search`` or ``parameter sweep``) to understand how the
```
Instead of "i.e." should this be "a.k.a." (also knowns as)

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1802,6 +1815,350 @@ def compute_FIM_full_factorial(
+            factorial_points = snake_traversal_grid_sampling(*design_values)
+        elif scheme_enum == SensitivityInitialization.nested_for_loop:
+            factorial_points = product(*design_values)
+
```
Does it make sense to split the function here? That way, we can do extensive testing on the results of `factorial_points` without running the rest of the function.

smondal13 replied:
Great suggestion, and I implemented that split: traversal generation now lives in a dedicated helper (_materialize_factorial_points).
Current tests already cover traversal behavior through the public API (including invalid scheme errors and explicit nested-for-loop ordering on a 2x2 grid).

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1802,6 +1815,350 @@ def compute_FIM_full_factorial(
+        ----------
+        model : DoE model, optional
+            The model to perform the full factorial exploration on. Default: None
+        design_vals : dict,
```
After reading this doc string, I do not fully appreciate all of options. I do not think we are looking at all of the edge cases with the tests.

smondal13 committed:
changed the docstring

smondal13 committed:
implemented suggestions from 9/23 pyomo dev meeting.

smondal13 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1802,6 +1815,376 @@ def compute_FIM_full_factorial(
+        rel_change: list = None,
+        n_design_points: int = None,
+        method="sequential",
+        return_df=True,
```
Since the method is printing the df depending on this flag, not returning it, does it make more sense to use `show_df` instead of `return_df`?

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

jsiirola committed:
Merge branch 'main' into sensitivity-initialization

blnicho committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
1. removed the argument `check_positive_definite` from `check_matrix()` in utils.py.
2. Raised error instead of warning if the DesignSpaceTraversal is wrong in doe.py

smondal13 committed:
Merge branch 'sensitivity-initialization' of github.com:smondal13/pyomo into sensitivity-initialization

blnicho committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

adowling2 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -233,3 +235,105 @@ def get_FIM_metrics(FIM):
+    # Set the index to the variable names if provided,
+    corr_df = (
+        pd.DataFrame(correlation_matrix, index=var_name, columns=var_name)
+        if var_name
```
What is this logic doing here?

smondal13 replied:
This ternary block returns a pd DataFrame if `var_name` is passed; otherwise, it returns a numpy array.

adowling2 reviewed (changes requested):
Please see my requested changes.

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Remove TODO and modify docstring.

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -233,3 +235,105 @@ def get_FIM_metrics(FIM):
     }
+
+
+def snake_traversal_grid_sampling(*array_like_args):
```
I want it to be a public function, that's why I have added it in `doe/utils.py`

smondal13 committed:
Refactor DesignOfExperiments documentation and improve correlation matrix computation

smondal13 committed:
Run black

smondal13 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -233,3 +235,105 @@ def get_FIM_metrics(FIM):
+    ----------
+    covariance_matrix : numpy.ndarray
+        2D array representing the covariance matrix.
+    var_name : list, optional
```
Added details in the docstring

smondal13 commented on pyomo/contrib/doe/doe.py:
Made the FIM metrics consistent in both `compute_FIM_full_factorial()` and `compute_FIM_factorial()`. Added design variable name error checking.

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

mrmundt commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1807,6 +1820,376 @@ def compute_FIM_full_factorial(
+        ----------
+        model : DoE model, optional
+            The model to perform the full factorial exploration on. Default: None
+        design_vals : dict,
```
I'll let @blnicho also chime in but my initial thoughts are:

The best general practice is to accept component objects / references, not names. It is convenient (for users) to optionally accept names and resolve them internally. So if it's possible, support both, with a clear canonical form.

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

jsiirola commented on pyomo/contrib/doe/tests/test_doe_errors.py:
```diff
@@ -219,8 +219,7 @@ def test_reactor_check_bad_prior_negative_eigenvalue(self):
             ValueError,
-            r"FIM provided is not positive definite. It has one or more "
-            r"negative eigenvalue\(s\) less than -{:.1e}".format(
+            r"Matrix provided is not positive definite. It has one or more negative eigenvalue\(s\) less than -{:.1e}".format(
```
This is a long line.  Can it be broken to stay within the 88-column limit?

jsiirola commented on pyomo/contrib/doe/utils.py:
```diff
@@ -266,3 +273,155 @@ def get_FIM_metrics(FIM):
+    """
+    Generates a multi-dimensional pattern for an arbitrary number of 1D array-like arguments.
+    This pattern is useful for generating patterns for sensitivity analysis when we want
+    to change one variable at a time. This function uses recursion and acts as a generator.
```
Please make sure these are wrapped within 88 columns.

jsiirola commented on pyomo/contrib/doe/utils.py:
```diff
@@ -91,41 +96,43 @@ def rescale_FIM(FIM, param_vals):
 
 
-def check_FIM(FIM):
+def check_matrix(mat):
```
This should be renamed to indicate what it is checking.  Further, it is not checking and returning True/False, it is asserting (see, e.g. the difference between `check_optimal_termination` and `assert_optimal_termination`).

Maybe rename to something like `assert_matrix_symmetric_positive_definite` or `assert_symmetric_positive_definite`?

smondal13 replied:
That's a very good suggestion. I have renamed the function name to make it more explicit.

jsiirola commented on pyomo/contrib/doe/utils.py:
```diff
@@ -91,41 +96,43 @@ def rescale_FIM(FIM, param_vals):
 
-    # Check if the FIM is positive definite
+    # Check if the matrix is positive definite
     if np.min(evals) < -_SMALL_TOLERANCE_DEFINITENESS:
```
Should these tolerances be passed as optional arguments that default to the global value if None?

smondal13 replied:
I think passing these tolerances as optional argument is a good idea. I have added them as optional arguments

jsiirola commented on pyomo/contrib/doe/utils.py:
```diff
@@ -266,3 +273,155 @@ def get_FIM_metrics(FIM):
+    -------
+    pandas.DataFrame/numpy.ndarray
+        If `var_name` is provided, returns a pandas DataFrame with the correlation matrix
+        and the specified variable names as both index and columns. If `var_name` is not
```
Please make sure these are wrapped <88 columns.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                else:
+                    raise ValueError(
+                        "Unexpected error in generating design values. Please check the"
+                        " input parameters."
```
Please make sure these are wrapped <88 columns.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                if k.name in design_vals.keys()
+            ]
+            # This ensures that the order of the design_values keys matches the order of the
+            # design_map_keys so that design_point can be constructed correctly in the loop.
```
Please make sure these are wrapped <88 columns.

smondal13 replied:
Done

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            Options are ``"snake_traversal"`` and ``"nested_for_loop"``.
+            If ``"snake_traversal"`` is used, the design variables will be initialized in a
+            snake-like pattern where only one design variables change at a time.
+            If ``"nested_for_loop"`` is used, the design variables will be initialized using a
```
Please make sure these are wrapped <88 columns.

jsiirola replied:
Same here: the quotes are not part of the string.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            )
+
+        if design_vals is not None:
+            # Check whether the design_ranges keys are in the experiment_inputs
```
`design_ranges` or `design_vals`?  This looks like a copy-paste error.

smondal13 replied:
Resolved

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            Mapping of design variable names (strings) to 1D array-like values:
+            ``{"var_name": [values...]}``. Keys must be a subset of
+            ``model.experiment_inputs`` names. Variables not included are fixed at
+            their model values.
```
Using names is fragile and best avoided.  Why are you using them?  This instance has an experiment, which uses a proper ComponentMap, so you should accept the same API here.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            match the number of design variables in ``model.experiment_inputs``.
+            If provided without ``abs_change``, ``abs_change`` is treated as zeros.
+        n_design_points : int, optional
+            Number of equally spaced points per design variable generated from
```
```suggestion
            Number of equally-spaced points per design variable generated from
```

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            If provided without ``abs_change``, ``abs_change`` is treated as zeros.
+        n_design_points : int, optional
+            Number of equally spaced points per design variable generated from
+            bounds via ``np.linspace``. Requires finite lower and upper bounds.
```
```suggestion
            bounds via :func:`np.linspace`. Requires finite lower and upper bounds.
```

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            bounds via ``np.linspace``. Requires finite lower and upper bounds.
+        method : str, optional
+            Method used for FIM computation. Options: ``"kaug"`` and
+            ``"sequential"``. Default: ``"sequential"``.
```
```suggestion
            Method used for FIM computation. Options: ``kaug`` and
            ``sequential``. Default: ``sequential``.
```
(Don't include the quotes, as that implies that the string contains quotes)

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+
+        # Group the mutually exclusive arguments for passing design values
+        # abs_change, and rel_change can be provided together or separately
+        num_des_provided = (
```
This implies you are computing the number of design points provided.  A less confusing variable might be `num_des_args_provided`

smondal13 replied:
yes, thank you. I have changed the name

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+        if num_des_provided > 1:
+            raise ValueError(
+                "Please provide only one of the following arguments: "
+                "`design_vals`, `n_design_points`, or `abs_change` and/or `rel_change`."
```
This is confusing.  I think you want to say, "design_vals, n_design_points, and abs_change/rel_change are mutually exclusive."

smondal13 replied:
Yes. I have updated the error message

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            raise ValueError(
+                "At least one of the following arguments: "
+                "`design_vals`, `n_design_points`, or `abs_change` and/or `rel_change`"
+                " must be provided."
```
This is confusing.  Something like "Missing required argument: specify one of design_vals, n_design_points, or abs_change/rel_change."?

smondal13 replied:
Yes. I have updated the error message

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            Relative change for each design variable. Used only when
+            ``design_vals`` and ``n_design_points`` are not provided. Length must
+            match the number of design variables in ``model.experiment_inputs``.
+            If provided without ``abs_change``, ``abs_change`` is treated as zeros.
```
Should these also accept a single float, that we copy to make an appropriately-sized list?

I also think that the names are misleading.  Perhaps `abs_step` and `rel_step` would be more clear?

smondal13 replied:
Yes. I think that would be better.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+
+                # if abs_change and/or rel_change is provided, generate design values
+                # using the formula:
+                # change_in_value = lower_bound * rel_change + abs_change
```
This formula seems suspect.  I would have expected it to be
```
change_in_value = (upper_bound - lower_bound) * rel_change + abs_change

smondal13 replied:
I did implement it as a relative change (step) of the lower bound. But using their range may be a better choice.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
 
         return self.fim_factorial_results
 
+    def compute_FIM_factorial(
```
This function is a beast and should be broken up into helper functions.  At the very least, pull the (probably overly complex) logic for generating the design points into it's own (private) function.

mrmundt replied:
+1 on this. This function is _huge_.

smondal13 replied:
@jsiirola Thanks for the suggestion. I agree and have refactored this in the latest update.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                        "`rel_change` must have the same length of "
+                        f"`{len(design_keys)}` as `design_keys`."
+                    )
+
```
The logic here is pretty convoluted.  You repeatedly test to see which "source" you are using for the design points.  I would recommend refactoring.

smondal13 replied:
I have refactored and added private methods.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                        f"{comp.name} does not have a lower or upper bound."
+                    )
+
+                if abs_change is None and rel_change is None:
```
Note that at this point you don't have to test both abs and rel: they are either both None or both non-None.

smondal13 replied:
I have refactored the code.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                    val = lb
+                    while val <= ub:
+                        des_val.append(val)
+                        val += del_val
```
```suggestion
                    des_val = list(range(lb, ub, del_val))
```

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                        2,
+                    ),
+                )
+            except:
```
Bare `except:` is potentially problematic, as it catching everything including both solver errors and generic programming errors.  At a minimum, you should include the exception in the log so that a user can get an idea as to why things failed.

smondal13 replied:
Added the error message inside the logger warning.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+                pseudo_A_opt,
+                E_opt,
+                ME_opt,
+            ) = compute_FIM_metrics(FIM)
```
This API should be re-thought: returning this many values is asking for a programming error (e.g., getting the order wrong).  I would strongly recommend changing the return value to a `namedtuple` (or a general class, if appropriate).

smondal13 replied:
I have converted this to a named tuple. I think this is better since the returned object should be immutable.

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2005,6 +2018,378 @@ def compute_FIM_full_factorial(
+            print("Success count:", success_count)
+            print("Failure count:", failure_count)
+            print("\n")
+            print(res_df)
```
We generally strongly discourage `print()` within Pyomo.  Instead, output should be channeled via loggers or global streams (like `sys.stdout`) that are easily replaceable.

smondal13 replied:
I have repaced the `print` with `sys.stdout`

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2081,23 +2474,63 @@ def draw_factorial_figure(
+        if not isinstance(fixed_design_variables, dict):
+            raise ValueError("``fixed_design_variables`` must be a dictionary.")
+        if any(not isinstance(name, str) for name in fixed_design_variables.keys()):
+            raise ValueError("Fixed design variable names must be strings.")
```
This kind of type checking is very "non-Pythonic" and something I would generally discourage.  Python encourages duck-typing: as long as the variable *behaves* like the desired type, then it should be OK.

smondal13 replied:
I have removed the strict type checks,

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 committed:
Apply suggestions from code review

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

smondal13 committed:
Enhance DesignOfExperiments API documentation and validation for design_vals input; add tests for component key normalization and error handling

smondal13 committed:
Refactor DesignOfExperiments parameters: rename abs_change/rel_change to abs_step/rel_step and update related logic; add test for factorial step formula using design span

smondal13 committed:
Update parameter names in DesignOfExperiments: change abs_change/rel_change to abs_step/rel_step in docstring and test case

smondal13 committed:
Update comments in DesignOfExperiments: rename change_in_value to step_change and update related comments for clarity

smondal13 committed:
Apply delta_stepp suggestion from code review

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

smondal13 committed:
Merge branch 'sensitivity-initialization' of github.com:smondal13/pyomo into sensitivity-initialization

smondal13 committed:
Refactor FIM metrics handling: Introduce named tuple for structured access and update related functions. Run black

smondal13 committed:
Replace check_matrix with assert_symmetric_positive_definite and enhance error handling. Update related tests for consistency.

smondal13 committed:
Add FactorialDesignMode enum and refactor compute_FIM_factorial method for improved design space handling

smondal13 committed:
Refactor factorial computation and enhance tests for design modes: Improve output formatting and add validation for design mode arguments in compute_FIM_factorial.

smondal13 committed:
Add test for compute_FIM_factorial with nested for loop order on 2x2 grid

smondal13 committed:
Merge branch 'main' into sensitivity-initialization

smondal13 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -233,3 +235,105 @@ def get_FIM_metrics(FIM):
+
+
+def compute_correlation_matrix(
+    covariance_matrix, var_name: list = None, significant_digits=3
```
Done.

smondal13 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -233,3 +235,105 @@ def get_FIM_metrics(FIM):
+    var_name : list, optional
+        List of variable names corresponding to the rows/columns of the covariance matrix.
+    significant_digits : int, optional
+        Number of significant digits to round the correlation matrix to. Default: 3.
```
Done

smondal13 commented on pyomo/contrib/doe/utils.py:
```diff
@@ -233,3 +235,105 @@ def get_FIM_metrics(FIM):
+
+    Parameters
+    ----------
+    covariance_matrix : numpy.ndarray
```
Good point. We now support square DataFrame input and have added explicit label checks: row labels must be unique, column labels must be unique, and row/column labels must match in the same order. If any of those fail, we raise a `ValueError` with a specific message.

smondal13 committed:
Enhance compute_correlation_matrix to validate DataFrame input: ensure unique row and column labels, and match between them. Add corresponding unit tests for validation.

smondal13 committed:
Run black

smondal13 committed:
Merge branch 'main' into sensitivity-initialization
